### PR TITLE
Noah van Ekdom Mid-Mod (2207 BE)

### PR DESCRIPTION
**Pull request by noahvanekdom**
Thought it went pretty well overall, I got a little in the weeds with trying to create test objects that lacked a tag for edge case testing, but I will review that over my lunch break.

### CODE REVIEW

File: spec/features/Welcome/index_spec.rb [hunk 1]
```diff
@@ -10,4 +10,12 @@
 
     expect(current_path).to eq("/quotes")
   end
+
+  it 'can return a list of up to 10 quotes based on search parameters' do
```
Comment (dcoleman21) [inline spec/features/Welcome/index_spec.rb:14]: Did you mean to have this test in your `quotes/index_spec.rb` :) 
The only thing that I would add is
```
expect(page).to have_content("Quote: You can't shake hands with a clenched fist")
```
The `Quote` part since it is also on the page. Particular I know lol

File: spec/features/quotes/index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      visit root_path
+      fill_in :keyword, with: 'I am'
+      click_on "Search for Quotes"
+    end
+    it "displays the total number of quotes" do
+      quote_count = QuotesFacade.searched_quotes("I am").count
+      expect(page).to have_content("Total Matching Quotes: #{quote_count}")
+    end
+
+    it "displays a list of the top ten quotes" do
```
Comment (dcoleman21) [inline spec/features/quotes/index_spec.rb:15]: What did you want here?

File: spec/features/quotes/index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    end
+
+    it "has information on each quote" do
```
Comment (dcoleman21) [inline spec/features/quotes/index_spec.rb:19]: What did you want here?

File: spec/features/quotes/index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    end
+  end
+
+  describe "sad paths" do
+
+    xit "if there are no tags associated with a quote, it does not display the tag category" do
```
Comment (dcoleman21) [inline spec/features/quotes/index_spec.rb:26]: Reach out to some of your peers to see how they thought about testing for no tags.

File: spec/features/quotes/index_spec.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
+require 'rails_helper'
+
+RSpec.describe 'Returning Quotes' do
```
Comment (dcoleman21) [inline spec/features/quotes/index_spec.rb:3]: The only thing that you are missing in your testing is to also display the keyword that was used in the search
```
As well as the keyword that's being used for the search
```

File: config/routes.rb [hunk 6]
```diff
@@ -1,4 +1,7 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   root 'welcome#index'
+
+  resources :quotes, only: [:index]
```
Comment (dcoleman21) [inline config/routes.rb:5]: Nice job using resources

File: app/views/quotes/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<h2> Here are your quote results: </h2>
```
Comment (dcoleman21) [inline app/views/quotes/index.html.erb:1]: This looks good. You could add an `else` to your `if` to check for no tags

File: app/services/quotes_service.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+class QuotesService
+  def self.conn
+    Faraday.new(url: 'https://api.quotable.io/')
```
Comment (dcoleman21) [inline app/services/quotes_service.rb:3]: Nice job making this helper method. I would remove the last `/` from the url. You have it on your `conn.get` url.

File: app/services/quotes_service.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+  end
+
+  def self.searched_quotes(keyword)
+    response = conn.get('/search/quotes', { query: keyword, limit: 150 })
```
Comment (dcoleman21) [inline app/services/quotes_service.rb:7]: what are you limiting here? I would probably limit this to what the story specifies which is 10. Nice job limiting here at the call.

File: app/poros/quote.rb [hunk 10]
```diff
@@ -0,0 +1,10 @@
+class Quote
+  attr_reader :id, :content, :author, :tags
+
+  def initialize(attributes)
+    @id = attributes[:_id]
```
Comment (dcoleman21) [inline app/poros/quote.rb:5]: When I ran your tests, line 5 failed with this error:
```
1) Quote PORO creates a quote object
     Failure/Error: @id = attributes[:_id]

     TypeError:
       no implicit conversion of Symbol into Integer
     # ./app/poros/quote.rb:5:in `initialize'
     # ./spec/poros/quote_spec.rb:8:in `new'
```
I don't even think that you need `:id`. It isn't one of the attributes that we need.

File: app/poros/quote.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+    @content = attributes[:content]
+    @author = attributes[:author]
+    @tags = attributes[:tags].join(", ")
```
Comment (dcoleman21) [inline app/poros/quote.rb:8]: Nice job getting the tags in a comma separated string vs an array.

File: app/facades/quotes_facade.rb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+class QuotesFacade
+  def self.searched_quotes(keyword)
```
Comment (dcoleman21) [inline app/facades/quotes_facade.rb:2]: This looks good for getting the quote attributes that we want but what about the `:totalCount`?? Here is a way that you can get both with a single call:
```
def self.searched_quotes(keyword)
    json = QuotesService.searched_quotes(keyword)
    @quotes = json[:results].map do |quote_data|
      Quote.new(quote_data)
    end
    [json[:totalCount], @quotes]
end
```
Now we can give our controller access to both of these as well

File: app/controllers/quotes_controller.rb [hunk 13]
```diff
@@ -0,0 +1,6 @@
+class QuotesController < ApplicationController
+  def index
```
Comment (dcoleman21) [inline app/controllers/quotes_controller.rb:2]: See my comments in your facade. If you implement passing those two things - `:totalCount and quotes`, your controller might look like:
```
def index
  @search_phrase = params[:query]
  @total_found, @quotes = QuotesFacade.searched_quotes(@search_phrase)
end
```
Now you have access to the total count and your quotes object that is holding the attributes you want. 

The `.first(10)` could be gotten rid of if you set your limit to 10 in your service.

**Review by dcoleman21 — COMMENTED** [review]
Overall you did pretty well. Make sure you are writing tests for everything that you implement. You did not get a Facade test written and you missed making sure that you included the keyword that was used to search. I would work on getting your testing more robust. Please see the comments/feedback that I left. Let us know if you have any question.
